junit-platform-engine/src/main/java/org/junit/platform/engine/support/store/NamespacedHierarchicalStore.java

```diff
@@ -0,0 +1,335 @@
+	private final CloseAction closeAction;
+
+	/**
+	 * Create a new store with the supplied parent.
+	 *
+	 * @param parentStore The parent store to use for lookups; may be
+	 *                    {@code null}.
+	 */
+	public NamespacedHierarchicalStore(NamespacedHierarchicalStore<N> parentStore) {
+		this(parentStore, null);
+	}
+
+	/**
+	 * Create a new store with the supplied parent and close action.
+	 *
+	 * @param parentStore The parent store to use for lookups; may be
+	 *                    {@code null}.
+	 * @param closeAction The action to be called for each stored value when
+	 *                    this store is closed.
+	 */
+	public NamespacedHierarchicalStore(NamespacedHierarchicalStore<N> parentStore, CloseAction closeAction) {
+		this.parentStore = parentStore;
+		this.closeAction = closeAction;
+	}
+
+	/**
+	 * If a close action is configured, it will be called with all successfully
+	 * stored values in reverse insertion order.
+	 */
+	@Override
+	public void close() {
+		if (closeAction == null) {
+			return;
+		}
+		ThrowableCollector throwableCollector = new ThrowableCollector(__ -> false);
+		storedValues.values().stream() //
+				.filter(storedValue -> storedValue.evaluateSafely() != null) //
+				.sorted(REVERSE_INSERT_ORDER) //
+				.map(StoredValue::evaluate) //
+				.forEach(value -> throwableCollector.execute(() -> closeAction.close(value)));
+		throwableCollector.assertEmpty();
+	}
+
+	/**
+	 * Get the value stored for the supplied namespace and key in this or the
+	 * parent store, if present.
+	 *
+	 * @return stored value; may be {@code null}
+	 */
+	public Object get(N namespace, Object key) {
+		StoredValue storedValue = getStoredValue(new CompositeKey<>(namespace, key));
+		return StoredValue.evaluateIfNotNull(storedValue);
+	}
+
+	/**
+	 * Get the value stored for the supplied namespace and key in this or the
+	 * parent store, if present, and cast it to the supplied required type.
+	 *
+	 * @return stored value; may be {@code null}
+	 * @throws NamespacedHierarchicalException if the stored value cannot be cast
+	 *                                         to the required type
+	 */
+	public <T> T get(N namespace, Object key, Class<T> requiredType) {
+		Object value = get(namespace, key);
+		return castToRequiredType(key, value, requiredType);
+	}
+
+	/**
+	 * Get the value stored for the supplied namespace and key in this or the
+	 * parent store, if present, or call the supplied function to compute it.
+	 *
+	 * @return stored value; may be {@code null}
+	 */
+	public <K, V> Object getOrComputeIfAbsent(N namespace, K key, Function<K, V> defaultCreator) {
+		CompositeKey<N> compositeKey = new CompositeKey<>(namespace, key);
+		StoredValue storedValue = getStoredValue(compositeKey);
+		if (storedValue == null) {
+			StoredValue newValue = storedValue(new MemoizingSupplier(() -> defaultCreator.apply(key)));
+			storedValue = Optional.ofNullable(storedValues.putIfAbsent(compositeKey, newValue)).orElse(newValue);
```

Comment:
Is this really necessary? The `ConcurrentHashMap.computeIfAbsent` already gives a guarantee of atomic execution.

Reply:
Which part are you referring to? The `MemoizingSupplier` or using `putIfAbsent`?

Reply:
Using `putIfAbsent` and creating the supplier outside, what advantage does this have over `storedValues.computeIfAbsent(compositeKey, key -> storedValue(new MemoizingSupplier(() -> defaultCreator.apply(key))))`

Reply:
Point taken. Changed in 791a755a94f5a3ed80cebac9322c317e0ccdb09f.